File: apps/android/app/src/main/kotlin/com/fnb/posterminal/ws/Protocol.kt
package com.fnb.posterminal.ws

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator
import kotlinx.serialization.json.JsonNames
import kotlinx.serialization.ExperimentalSerializationApi

/**
 * Wire-format messages exchanged with the API server. Mirrors
 * `apps/api/src/pos-terminal/protocol.ts` exactly — keep both in sync.
 *
 * The server uses Zod's `discriminatedUnion` on `type`; kotlinx.serialization
 * lines up with that using `@JsonClassDiscriminator("type")` on the sealed
 * root and `@SerialName` on each variant.
 */

@OptIn(ExperimentalSerializationApi::class)
@Serializable
@JsonClassDiscriminator("type")
sealed interface ServerMessage {

    @Serializable
    @kotlinx.serialization.SerialName("ready")
    data class Ready(
        val locationName: String,
        val tenantName: String,
    ) : ServerMessage

    @Serializable
    @kotlinx.serialization.SerialName("pong")
    object Pong : ServerMessage

    @Serializable
    @kotlinx.serialization.SerialName("payment_request")
    data class PaymentRequest(
        val intentId: String,
        /**
         * "kiosk_order" — customer ordered through the kiosk web app and is
         * paying now; on approve, items fire to the kitchen.
         * "staff_ticket" — cashier rang up a ticket at the POS and the
         * kitchen already has it; the customer is paying at the counter.
         * Drives the messaging on the result screen.
         */
        val kind: String? = null,
        val amountCents: Long,
        val currency: String,
        val shortNumber: Int,
        val customerName: String,
        val tableLabel: String?,
        val items: List<LineItem>,
    ) : ServerMessage

    @Serializable
    @kotlinx.serialization.SerialName("payment_cancel")
    data class PaymentCancel(
        val intentId: String,
        val reason: String,
    ) : ServerMessage
}

@Serializable
data class LineItem(
    val name: String,
    val qty: Int,
    val lineTotalCents: Long,
    val modifiers: List<String> = emptyList(),
)

@OptIn(ExperimentalSerializationApi::class)
@Serializable
@JsonClassDiscriminator("type")
sealed interface ClientMessage {

    @Serializable
    @kotlinx.serialization.SerialName("payment_result")
    data class PaymentResult(
        val intentId: String,
        val status: PaymentResultStatus,
    ) : ClientMessage

    @Serializable
    @kotlinx.serialization.SerialName("ping")
    object Ping : ClientMessage
}

@Serializable
enum class PaymentResultStatus {
    APPROVED,
    DECLINED,
    CANCELLED,
}

File: apps/api/src/pos-terminal/protocol.ts
import { z } from 'zod';

/**
 * Wire-format messages exchanged with the Android POS terminal.
 *
 * Inbound (terminal → server):
 *   - payment_result: card-read outcome for a specific intent
 *   - ping:           heartbeat
 *
 * Outbound (server → terminal):
 *   - ready:           handshake ack after the terminal connects
 *   - payment_request: customer kiosk asked for a card-read on this terminal
 *   - payment_cancel:  in-flight intent was voided server-side (auto-expire,
 *                      explicit cancel from kiosk, etc.)
 *   - pong:            heartbeat ack
 */

export const inboundMessageSchema = z.discriminatedUnion('type', [
  z.object({
    type: z.literal('payment_result'),
    intentId: z.string().uuid(),
    status: z.enum(['APPROVED', 'DECLINED', 'CANCELLED']),
  }),
  z.object({ type: z.literal('ping') }),
]);
export type InboundMessage = z.infer<typeof inboundMessageSchema>;

export interface PaymentRequestPayload {
  intentId: string;
  amountCents: number;
  currency: string;
  shortNumber: number;
  customerName: string;
  tableLabel: string | null;
  items: Array<{
    name: string;
    qty: number;
    lineTotalCents: number;
    modifiers: string[];
  }>;
}

export type OutboundMessage =
  | { type: 'ready'; locationName: string; tenantName: string }
  | { type: 'pong' }
  | ({ type: 'payment_request' } & PaymentRequestPayload)
  | { type: 'payment_cancel'; intentId: string; reason: string };

export function encode(msg: OutboundMessage): string {
  return JSON.stringify(msg);
}
